src/v/cluster/metadata_dissemination_service.cc

```diff
@@ -270,19 +270,13 @@ void metadata_dissemination_service::collect_pending_updates() {
             // Topic metadata is not there anymore, partition was removed
```

Comment:
Was trying to figure this out too, I think the concern is that nodes decide where to send updates based on their knowledge of placement at the next tick after apply_leadership_notification happens, but there's no guarantee that that's up to date at the time our updates are dispatched.

Not sure if that's actually the case that came up as a bug in practice?  @mmaslankaprv ?

Reply:
The problem with non_overlapping nodes was related with the fact that metadata dissemination happens asynchronously to partition movement process. We may miss notification to a node if partition movement happens fast enough. Metadata which we use to disseminate leadership are updated before the actual raft configuration update mechanism is triggered. 

I agree that we should optimize it in future, we will have cluster health monitor shortly, and then we may go back with the previous approach.

Reply:
i understand there is a race, and that by updating _all_ nodes, the race still exists but is rendered benign because the update is effectively forced to overwrite the stale state that resulted from the race. but i don't think we've explained what the race is in a way that we can figure out issues in the future.

here is my attempt, but i'm still not able to describe a scenario that causes a problem:

all nodes = {a, b, c, d, e}
raft group replica set = {a, b, c}
leader = a

when leadership changes in the raft group, updates are sent to {d, e} and not {a, b, c} because {a, b,c} already know the leadership of the raft group because they are members of the replica set.

> Metadata which we use to disseminate leadership are updated before the actual raft configuration update mechanism is triggered.

the race is that the replica set is being changed from {a, b, c} to {a, b, d}. presumably _Metadata which we use to disseminate leadership are updated before_ is referring to an update in which current leadership information is broadcast to {c, e} (because c is now not part of the raft group).

so if dissemination framework has the update replica set then any subsequent change in raft should trigger updates to the correct non-overlapping group?

Reply:
I was trying to find an example for this and i can not. i will run the tests without dissemination for each and every node to make sure i understand the corner case.

Reply:
I've removed two commits changing the metadata updates to broadcast. I couldn't find any scenarios in which the dissemination might fail.